edx_prefectutils/amplitude.py

```diff
@@ -0,0 +1,347 @@
+    don't exceed batch_size events.  Then, it will batch any segments still larger than batch_size. The final product
+    should look like an iterable of iterable of iterable of events. I.e.:
+
+    - First level loops over user/device segment (each of these will be allocated to a different worker thread).
+    - Second level loops over batch_size batches of events (this is a requirement imposed by the Amplitude API).
+    - Third level loops over individual events.
+
+    Memory Efficiency: In practice, this function won't fetch all events for the current processing segment at once;
+    instead it will fetch small amounts until there are enough events to go around for all the worker threads.  The
+    best-case-scenario memory consumption would be about N_worker_threads * batch_size * event_size, which is tiny.
+    However, the worse-case-scenario is that somehow a large proportion of the events in this processing segment
+    corresponds to a single user, in which case the memory consumption would equal that of loading the entire processing
+    segment into memory (multiple gigabytes).  We could address memory consumption by adding even more complexity (and
+    possibly bugs) to the code, but this is an exceedingly rare scenario that isn't worth optimizing for.
+
+    Args:
+        cursor (Snowflake cursor): Cursor of an executed query from which to fetch results.
+        batch_size (int): Maximum number of events to include in a single Amplitude API call.  This cannot be greater
+            than 1000 since that is currently the maximum imposed by Amplitude.
+        fetch_size (int): The "small" number of events to fetch from snowflake each time we need to ask for more.
+
+    Returns:
+        iterable of iterable of iterable of events.
+    """
+    accumulated_events = LazyResultFetcher(cursor, fetch_size=fetch_size)
+    last_idx_for_worker = None
+    while True:
+        if accumulated_events.actual_length() <= batch_size:
+            # All we have left is one or more segments that completely fit into one batch, so send it all!
+            last_idx_for_worker = len(accumulated_events)
+        else:
+            if accumulated_events[0].user_device == accumulated_events[batch_size - 1].user_device:
+                # We have a segment which is larger than a single bucket.  We should find the last event for that
+                # segment and send the entire segment to a worker.
+                target_user_device = accumulated_events[0].user_device
+                idx = batch_size - 1
+                while True:
+                    # Is there a more efficient way than indexing one event at a time?  Yes, but simple iteration is not
+                    # our performance bottleneck, and anything more complicated can introduce bugs.
+                    idx += 1
+                    try:
+                        if accumulated_events[idx].user_device != target_user_device:
+                            last_idx_for_worker = idx
```

Comment:
So last_idx is not inclusive.   Makes sense for slicing.   But I wonder if "last" is misleading.   Maybe next_idx?  Nah.  Hmm.    Fine...

Reply:
Yea, it's a bit misleading.  end_idx?